compiler/src/dotty/tools/dotc/typer/Inferencing.scala

```diff
@@ -95,8 +96,13 @@ object Inferencing {
             if (minimize) instantiate(tvar, fromBelow = true)
             else toMaximize = true
           }
-          foldOver(x, tvar)
         }
+        foldOver(x && forceApplies,
```

Comment:
I think this is a good fix, but the logic can be kept closer to the original one, and I think that's both shorter and clearer. How about:

        force.appliesTo(tvar) && {
           val direction = instDirection(tvar.origin)
            if (direction != 0) {
              //if (direction > 0) println(s"inst $tvar dir = up")
             instantiate(tvar, direction < 0)
           }
           else {
             val minimize =
               force.minimizeAll ||
               variance >= 0 && !(
                 force == ForceDegree.noBottom &&
                 defn.isBottomType(ctx.typeComparer.approximation(tvar.origin, fromBelow = true)))
              if (minimize) instantiate(tvar, fromBelow = true)
              else toMaximize = true
            }
            tvar.isInstantiated || foldOver(x, ctx.typerState.constraint.entry(tvar.origin))    
         }